### ISSUE TITLE
Numero Uno

### ISSUE THREAD

**Issue by agzam** [craftinghammers/spacehammer#1]
It's just a placeholder for a preliminary draft of the list of things to consider.

This is the beginning phase of designing the next iteration of agzam/spacehammer.

## The issues I personally face with the current version:

- There is a memory leak I can't figure out. I suspect it's an upstream issue, but we still need to find a workaround. Perhaps, while redesigning the whole app, we'll find the cause. Basically, for no apparent reason, Spacehammer's main menu freezes or doesn't show up at all; Activity Monitor shows a sudden increase in memory usage. It can jump from about a couple of hundred MBs to several gigs 😱 in second. And then the only way is to forcefully kill and restart Hammerspoon.

- I would absolutely love to have proper tiling WM support. 

- I have never figured out how to set up a proper Fennel REPL, but I would love to write forms and evaluate them without leaving my editor.

## What I think would be great to have

- An easier way to install things - Nix, Homebrew, or luarocks, etc.

- A simpler way of adding menu items would be fantastic. It would be nice to have a universal macro that would let you easily define and extend menus and items.

- Local bindings support. It would be great if we had a dedicated key that invokes a local menu, specifically made for the app, or maybe even different classes of apps. For example, I imagine apps that have "tabs" - Browser, Terminal, Finder, etc., could add some similar features through the local menu. We have a way of doing local bindings today, but it's not as simple and straightforward to use, possibly due to the general difficulty of making and configuring nested menus. Would be nice to be able to easily "hook" up any app and make it for example always start with a specific screen-rect coords. 

- I don't know if we need to revisit our FSM implementation, but some optimizations wouldn't hurt, I suppose.

- End-to-end testing is something I've always wanted, and I'm not sure if that's even attainable via practical means, especially unsure if we can do it via GH Actions.

- In general, a better way of customizing things would be great. We should aim to have a very tiny core and make things extremely extensible.

- Some consensus around the "core" library is needed. Fennel doesn't have an official core library; either adopting something like andreyorst/fennel-cljlib or borrowing things from other sources could potentially work.

**Comment by jaidetree** [comment]
> It's just a placeholder for a preliminary draft of the list of things to consider.
> 
> This is the beginning phase of designing the next iteration of agzam/spacehammer.
> 
> ## The issues I personally face with the current version:
> * There is a memory leak I can't figure out. I suspect it's an upstream issue, but we still need to find a workaround. Perhaps, while redesigning the whole app, we'll find the cause. Basically, for no apparent reason, Spacehammer's main menu freezes or doesn't show up at all; Activity Monitor shows a sudden increase in memory usage. It can jump from about a couple of hundred MBs to several gigs 😱 in second. And then the only way is to forcefully kill and restart 

Haven't run into that one yet but will keep an eye out.

> * I would absolutely love to have proper tiling WM support.

That's one of the reasons I'm pushing for spoonification, so people can try things like https://github.com/mogenson/PaperWM.spoon. It will also make it easier to build our own as an experimental spoon people can opt into.

> * I have never figured out how to set up a proper Fennel REPL, but I would love to write forms and evaluate them without leaving my editor.

That's probably worth re-visiting anyway given that Fennel as a more fleshed out repl in recent releases.
 
> ## What I think would be great to have
> * An easier way to install things - Nix, Homebrew, or luarocks, etc.

Nix would definitely be cool and seems doable, might even be easier once the spoonification is complete.

> * A simpler way of adding menu items would be fantastic. It would be nice to have a universal macro that would let you easily define and extend menus and items.

@ambirdsall as lead and @grazfather and I for support have been drafting an API based version to use things like `(map! :leader :a max-window)` and setting extra leaders. This is underway and we have been using a thread in the doom-emacs discord.

> * Local bindings support. It would be great if we had a dedicated key that invokes a local menu, specifically made for the app, or maybe even different classes of apps. For example, I imagine apps that have "tabs" - Browser, Terminal, Finder, etc., could add some similar features through the local menu. We have a way of doing local bindings today, but it's not as simple and straightforward to use, possibly due to the general difficulty of making and configuring nested menus. Would be nice to be able to easily "hook" up any app and make it for example always start with a specific screen-rect coords.

I think this is possible with the upcoming API changes mentioned just above.

> * I don't know if we need to revisit our FSM implementation, but some optimizations wouldn't hurt, I suppose.

Currently rewriting the FSM library again, its intended to handle effects and validate transitions and states to enforce correctness at runtime.

> * End-to-end testing is something I've always wanted, and I'm not sure if that's even attainable via practical means, especially unsure if we can do it via GH Actions.

I'm not sure full end-to-end testing is doable atm since I'm unaware of a system that can detect if windows render and move to expected coordinates, jumping focus, shortcuts, etc... but we could at least take the unit tests and might be able to get those in a GH action. From what I understand github does support a mac os container for actions. If we can get it to install hammerspoon and spacehammer it could run the tests.

> * In general, a better way of customizing things would be great. We should aim to have a very tiny core and make things extremely extensible.

This was also part of the API plans we've been discussing and are slowly working towards. The goal is to slim down the default config to the main concepts, comment out some extras, and add more detailed docs as comments in the default config so it's simpler and easier to jump in.

> * Some consensus around the "core" library is needed. Fennel doesn't have an official core library; either adopting something like andreyorst/fennel-cljlib or borrowing things from other sources could potentially work.

I'll let others jump in on this one. Personally, I'd rather keep iterating on the functional library we started drafting than fennel-cljlib. It's a cool library but it's also very hefty trying to make fennel into something it's not: clojure. I love clojure but I think fennel's commitment to not adding overhead does provide a lot of value here. Again though, open to others who want to weigh in.

Anyway hopefully this is helpful.
